File: Algorithms/0915_Partition_Array_into_Disjoint_Intervals/Python/Partition_Array_into_Disjoint_Intervals_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def partitionDisjoint(self, nums):
        length = len(nums)
        if length == 2: return 1
        left_max, right_min = [-1 for _ in range(length)], [-1 for _ in range(length)]
        left_max[0] = nums[0]
        right_min[-1] = nums[-1]

        for i in range(1, length):
            left_max[i] = max(nums[i], left_max[i - 1])

        for i in range(length - 2, -1, -1):
            right_min[i] = min(nums[i], right_min[i + 1])

        for i in range(length - 1):
            if left_max[i] <= right_min[i + 1]: return i + 1

        return -1

File: Algorithms/0915_Partition_Array_into_Disjoint_Intervals/Java/PartitionArrayintoDisjointIntervalsSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int partitionDisjoint(int[] nums) {
        if (nums.length == 2) return 1;
        int[] cache = new int[nums.length];
        Arrays.fill(cache, -1);
        int res = 0;
        cache[0] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            cache[i] = Math.max(nums[i], cache[i - 1]);
        }
        int rightMin = nums[nums.length - 1];
        for (int i = nums.length - 2; i >= 0; i--) {
            if (rightMin >= cache[i]) res = i;
            rightMin = Math.min(rightMin, nums[i]);
        }
        return res + 1;
    }
}




